Key export

OrestisAlpos opened the pull request:
Added functionality for key export.
I mimicked the logic for key import, hence created two new projects `key-export-common` and `key-export-client`. The former is used by the latter and by the signers.
I renamed the existing `common` and `client` folders to `key-import-common` and `key-import-client`, respectively. The encryption code still remains in `key-import-common`.

Commit:
rename common and client to key-import common and key-import client, reps;  add key-export code

Commit:
fix imports, remove unused feature

Commit:
clippy suggestions

Commit:
fix import

Commit:
update github workflows to match new directory names + add wasm build for key-export client

Commit:
cargo mistakenly thinks KeyExportError::new() is unused here

OrestisAlpos commented:
I also updated `.github/workflows` accordingly.

survived commented on key-export-common/Cargo.toml:
```diff
@@ -0,0 +1,21 @@
+generic-ec-zkp = { git = "https://github.com/dfns-labs/generic-ec", branch = "m", default-features = false, features = ["serde", "alloc"] }
+
+# We are using the encryption scheme defined in key-import repo:
+dfns-key-import-common = { path = "../key-import-common" }
```
We should rather move the encryption key to a separate package. Having key export depending on key import is odd.

OrestisAlpos replied:
Will do. I will message you for some details.

survived commented on key-export-client/tests/export-request.rs:
```diff
@@ -0,0 +1,14 @@
+use dfns_key_export_client::KeyExportContext;
+
+// Not a real test, just prints out the key-export request that is to be sent to Dfns API.
```
If it's not a real test, then it probably shouldn't be in the tests folder? E.g. you can move it to examples folder.

survived commented on key-export-client/tests/export-request.rs:
```diff
@@ -0,0 +1,14 @@
+        .unwrap();
+    println!("{:?}", req);
+    // let req_parsed: KeyExportRequest = serde_json::from_str(&req).unwrap();
+    // println!("{}", hex::encode(req_parsed.encryption_key.to_bytes()));
```
Did you forget to delete? Or otherwise, why did you leave this comment here?

survived commented on key-export-common/src/lib.rs:
```diff
@@ -0,0 +1,198 @@
+#[derive(Clone, Debug, serde::Serialize, serde::Deserialize)]
+pub struct KeySharePlaintextNoGen {
+    /// The index (evaluation point)
+    // pub index: NonZero<Scalar<Secp256k1>>,
```
If this field is not needed, this needs to be deleted

survived commented on key-export-common/src/lib.rs:
```diff
@@ -0,0 +1,198 @@
+
+/// Format of a decrypted key share
+#[derive(Clone, Debug, serde::Serialize, serde::Deserialize)]
+pub struct KeySharePlaintextNoGen {
```
What's difference between `KeySharePlaintext` and `KeySharePlaintextNoGen`?

OrestisAlpos replied:
I am deleting `KeySharePlaintextNoGen`, I only wrote it to debug an error and I forgot it.

survived commented on key-export-common/src/lib.rs:
```diff
@@ -0,0 +1,198 @@
+    /// Signers's key share.
+    /// It is an encrypted `dfns_key_export_common::KeySharePlaintext`.
+    /// Ciphertext and plaintext are in format defined by `dfns-key-import-common`
+    /// library. See https://github.com/dfns-labs/trusted-dealer/
```
You need to put a new line to separate a header from the body of the comment. Also the link can be prettified

```suggestion
    /// Signers's key share.
    ///
    /// It is an encrypted `dfns_key_export_common::KeySharePlaintext`.
    /// Ciphertext and plaintext are in format defined by `dfns-key-import-common`
    /// library. See [here](https://github.com/dfns-labs/trusted-dealer/).
```

survived commented on key-export-common/tests/interpolate-from-shares.rs:
```diff
@@ -0,0 +1,73 @@
+    let public_key = public_key.to_bytes(true).to_vec();
+
+    let shares = split_secret_key(&mut rng, t, n, &secret_key).unwrap();
+    // println!("{:?}", serde_json::to_string(&shares[0]));
```
It's better to delete the commented pieces of code as they generally do not improve readability

survived commented on key-export-common/tests/interpolate-from-shares.rs:
```diff
@@ -0,0 +1,73 @@
+    let res = interpolate_secret_key::<E>(&shares[..2], &public_key);
+    assert!(res.is_err());
+    assert!(matches!(
+        res.expect_err(""),
```
If there's no message, `.unwrap_err()` should be used instead

survived commented on key-export-client/src/lib.rs:
```diff
@@ -0,0 +1,177 @@
+#[cfg(target_arch = "wasm32")]
+type ErrorType = JsError;
+#[cfg(not(target_arch = "wasm32"))]
+type ErrorType = KeyExportError;
```
Why do you keep support for non-wasm targets? Do we need the client code to be compiled to different targets?

OrestisAlpos replied:
Yes, I am using it in the [integration tests](https://github.com/dfnsco/signers/blob/62abba66bc39fb8053b5d9f75123ded1f4d9cbb5/tests/tests/integration.rs#L959)

survived replied:
Originally, the reason why I made separation between key-import-client and key-import-common is that common is used by both signer (including the tests) and client code. In key-import, the client is wasm only. If client code can be invoked by the signer directly, then why do we have the `key-export-common`?

OrestisAlpos replied:
There is still a logical separation, e.g. there is no wasm dependency in the `key-export-common`,  there are generics over an elliptic curve E in `key-export-common` and the client calls a specific version of them depending on the exported key type.
Most importantly, `key-export-client` is used only in the tests of signers. I just decided to write an end-to-end test in the end, that's how I discovered the serialization bug for example.
I can still merge them into one with different modules if you think it makes more sense.

survived replied:
I guess it does make sense to keep separation

survived commented on key-export-client/src/lib.rs:
```diff
@@ -0,0 +1,177 @@
+            (_, _) => {
+                return Err(new_error("the combination of protocol and curve for this key is not supported for key export"));
+            } // ed25519_dalek::PublicKey::from_bytes(public_key).context("invalid public key")?;
+              // Point::from_bytes(&response.public_key).context("invalid public key")?
```
Please remove all the commented out code

Commit:
tidy up: remove some comments and rename test to example

Commit:
add a test

Commit:
update generic-ec dependency import

survived commented on key-export-common/src/lib.rs:
```diff
@@ -0,0 +1,198 @@
+///
+/// `key_shares` is a vector of serialized KeySharePlaintext<E>
+/// 'public_key` is a serialized Point<E>
+pub fn interpolate_secret_key<E: Curve>(
```
Missing empty line between header and main body of the doc comment

I advice to do `cargo doc --open` to check that all documentation is displayed as intended

```suggestion
/// Internal function to perform interpolation.
///
/// In the end tt verifies the computed key against the provided
/// public key and returns an error if it doesn't match.
///
/// `key_shares` is a vector of serialized KeySharePlaintext<E>
/// 'public_key` is a serialized Point<E>
pub fn interpolate_secret_key<E: Curve>(
```

survived commented on key-export-common/src/lib.rs:
```diff
@@ -0,0 +1,198 @@
+pub fn interpolate_secret_key<E: Curve>(
+    key_shares_bytes: &[Vec<u8>],
+    public_key_bytes: &[u8],
+) -> Result<SecretScalar<E>, InterpolateKeyError> {
```
I think this function should rather accept already parsed `public_key` and `key_shares`, or it should be called `parse_and_interpolate_secret_key`

survived commented on key-export-common/src/lib.rs:
```diff
@@ -0,0 +1,198 @@
+///
+/// `key_shares` is a vector of serialized KeySharePlaintext<E>
+/// 'public_key` is a serialized Point<E>
+pub fn interpolate_secret_key<E: Curve>(
```
Is this function being used in the signers code?

OrestisAlpos replied:
It was not, and i refactored it a bit. It is now in `key-export-client` and I made it get the parsed shares. Does look better :D

survived commented on key-export-common/src/lib.rs:
```diff
@@ -0,0 +1,198 @@
+    let shares = key_shares
+        .iter()
+        .map(|s| (s.secret_share.clone()))
+        .collect::<Vec<SecretScalar<_>>>();
```
I don't think you actually need to collect this iter into the vec. From what I see, it's being used only once in `.zip(shares)`, you can feed an iter directly into this

survived commented on key-export-client/src/lib.rs:
```diff
@@ -0,0 +1,177 @@
+
+    /// Parse the response from Dfns API and recover the secret key.
+    /// The functions returns the secret key in big-endian format.
+    pub fn recover_secret_key(&self, response: String) -> Result<Vec<u8>, ErrorType> {
```
Should it rather return a `SecretKey` that provides different options for its serialization? E.g.

```rust
impl SecretKey {
    pub fn to_be_bytes(&self) -> Vec<u8>;
    pub fn to_le_bytes(&self) -> Vec<u8>;
}
```

OrestisAlpos replied:
We could choose that, but, according to the engineering doc, we only want the secret key in be_bytes.

survived replied:
I'm just thinking that there might be other formats of secret key we might want to support as an output in the future

survived replied:
So we can only expose `to_be_bytes` method now, and we'll have freedom to add another method, if we need to, in the future without breaking library API compatibility

OrestisAlpos replied:
Ok that's a good idea. I will change it.

Commit:
extract encryption to separate crate

Commit:
move interpolate_secret_key() from key-export-common to key-export-client

Commit:
remove unnecessary import

Commit:
re-export encryption

Commit:
forgotten re-export

Commit:
better cargo doc

Commit:
remove module from crate encryption

Commit:
remove pub from version

Commit:
pass feature down to reexported package

Commit:
return private key as a dedicated type

Commit:
fix to_bytes_be(), make it method of SecretKey

Commit:
move encryption to separate module

Commit:
move version logic from key-import-common to core

Commit:
change VersionGuard to use const generics

Commit:
add version guard to key export

Commit:
reexport encryption and version

Commit:
change serialize() of EncryptionKey to base64

survived commented on core/Cargo.toml:
```diff
@@ -9,7 +9,7 @@ generic-ec = { git = "https://github.com/dfns-labs/generic-ec", branch = "m", de
 rand_core = { version = "0.6", default-features = false }
 serde = "1"
-hex = { version = "0.4", default-features = false, features = ["alloc", "serde"]}
+base64 = "0.21.4"
```
Generally, you don't need to specify a patch version

```suggestion
base64 = "0.21"
```

survived commented on core/src/version.rs:
```diff
@@ -0,0 +1,70 @@
+                "you seem to be using old version of the library: version field of serialized \
+                data (v{version}) doesn't match version supported by this library (v{})",
+                V
+            )))
```
```suggestion
            Err(<D::Error as serde::de::Error>::custom(&alloc::format!(
                "you seem to be using old version of the library: version field of serialized \
                data (v{version}) doesn't match version supported by this library (v{V})"
            )))
```

survived commented on key-export-client/src/lib.rs:
```diff
@@ -0,0 +1,269 @@
+    /// export the key of the wallet with the given `wallet_id`.
+    ///
+    /// Throws `Error` in case of failure.
+    pub fn build_key_export_request(&self, wallet_id: String) -> Result<String, ErrorType> {
```
As I understood from the last call with Thibault, we don't need to accept `wallet_id`, client code will populate this and other fields

survived replied:
What about returning a `JsValue` we discussed? Or do you want to do that in a separate PR?

OrestisAlpos replied:
About `wallet-id`: We discussed about the client code populating fields on the key-import request. The key-export specifications still have the wallet-id as input to WASM. Let me ask Thibault.

I will change it to `JsValue` in this PR.

survived replied:
I think it's just outdated

survived commented on key-export-client/src/lib.rs:
```diff
@@ -0,0 +1,269 @@
+    }
+}
+
+/// Parse a collection of decrepted shares as a vector of``KeySharePlaintext<E>``.
```
```suggestion
/// Parse a collection of decrypted shares as a vector of `KeySharePlaintext<E>`.
```

survived commented on key-export-client/Cargo.toml:
```diff
@@ -0,0 +1,22 @@
+generic-ec = { git = "https://github.com/dfns-labs/generic-ec", branch = "m", default-features = false, features = ["serde", "curve-secp256k1", "wasm"] }
+generic-ec-zkp = { git = "https://github.com/dfns-labs/generic-ec", branch = "m", default-features = false, features = ["serde", "alloc"] }
+
+wasm-bindgen = "0.2.84"
```
Shouldn't it be an optional dependency that's only enabled for wasm target?

survived replied:
It should be:
```toml
[target.'cfg(target_arch = "wasm32")'.dependencies]
wasm-bindgen = "0.2"
```

OrestisAlpos committed:
Update core/src/version.rs

more concise in format!()

Co-authored-by: Denis Varlakov <denis@dfns.co>

OrestisAlpos committed:
Update key-export-client/src/lib.rs

fix typo

Co-authored-by: Denis Varlakov <denis@dfns.co>

Commit:
dependencies: do not specify patch, add wasm conditionally

Commit:
change return type of build_key_export_request()

Commit:
remove wallet_id from input to the WASM function

OrestisAlpos commented:
Updated PR: Accepted suggestions from Denis, removed the wallet_id from input to the WASM function, changed the return type of `build_key_export_request()`.
For the last one, I wanted to keep the support for compiling into wasm and non-wams targets (I am using the non-wasm, for example, in the `examples` of `key-export-client` and in the integration tests of signers). Let me know if the solution with the two modules `utils_wasm32` and `utils_non_wasm32` is very clean.

survived commented on key-export-client/src/lib.rs:
```diff
@@ -8,12 +8,22 @@
+#[cfg(not(target_arch = "wasm32"))]
+use utils_non_wasm32 as utils;
+#[cfg(target_arch = "wasm32")]
+use utils_wasm32 as utils;
```
The module should rather be called `types` as it basically defines request and error types.

OrestisAlpos replied:
I was debating with that as well. I will accept your suggestion.

survived commented on key-export-client/src/lib.rs:
```diff
@@ -8,12 +8,22 @@
+#[cfg(not(target_arch = "wasm32"))]
+use utils_non_wasm32 as utils;
+#[cfg(target_arch = "wasm32")]
+use utils_wasm32 as utils;
```
There's a more Rust-ish way to do that. (assuming we've renamed `utils` into `types`) You can create a folder `types` move `types_wasm32.rs` into `types/wasm32.rs` and `types_non_wasm32.rs` into `types/others.rs`.

Then you can write:
```suggestion
#[cfg_attr(target_arch = "wasm32", path = "types/wasm32.rs")]
#[cfg_attr(not(target_arch = "wasm32"), path = "types/others.rs")]
mod types;
```

OrestisAlpos replied:
Oh yes, much better!

survived commented on key-export-client/Cargo.toml:
```diff
@@ -16,8 +16,10 @@ generic-ec-zkp = { git = "https://github.com/dfns-labs/generic-ec", branch = "m"
+#[target.'cfg(target_arch = "wasm32")'.dependencies]
 wasm-bindgen = "0.2"
+serde = { version = "1.0", features = ["derive"] }
+serde-wasm-bindgen = "0.6"
```
I don't see why you commented out `[target.'cfg(target_arch = "wasm32")'.dependencies]`. It should be

```toml
[target.'cfg(not(target_arch = "wasm32"))'.dependencies]
serde = { version = "1.0", features = ["derive"] }

[target.'cfg(target_arch = "wasm32")'.dependencies]
wasm-bindgen = "0.2"
serde-wasm-bindgen = "0.6"
```

OrestisAlpos replied:
Ah no sorry, we do not need to import serde at all here actually

survived commented on key-export-client/src/lib.rs:
```diff
@@ -0,0 +1,245 @@
+}];
+
+/// Secret key to be exported
+#[cfg_attr(target_arch = "wasm32", wasm_bindgen::prelude::wasm_bindgen)]
```
I actually liked more when you had `use wasm_bindgen::prelude::*`

Commit:
refactor types mod

Commit:
remove comment from cargo.toml

survived commented on key-export-client/src/lib.rs:
```diff
@@ -84,20 +81,20 @@ impl KeyExportContext {
     ///
     /// Throws `Error` in case of failure.
-    pub fn build_key_export_request(&self) -> Result<utils::RequestType, utils::ErrorType> {
+    pub fn build_key_export_request(&self) -> Result<types::RequestType, types::ErrorType> {
```
Since the module is called `types`, and you prefix it everywhere, maybe it makes sense to remove `Type` suffix? I.e.
```suggestion
    pub fn build_key_export_request(&self) -> Result<types::Request, types::Error> {
```

survived commented on core/src/lib.rs:
```diff
@@ -0,0 +1,12 @@
+extern crate alloc;
+
+/// Version number, ensures that server and client are compatible
+const VERSION: u8 = 1;
```
If it's only used in `encryption`, shouldn't it be moved in there? And update the comment to reflect that it's a version of encryption scheme

survived commented on key-export-client/tests/interpolate_secret_key.rs:
```diff
@@ -0,0 +1,95 @@
+        secret_shares
+    };
+
+    let shares = (1..n + 1)
```
Not critical, but you can use 
```suggestion
    let shares = (1..=n)
```

survived reviewed (commented):
A few more suggestions!

Commit:
more suggestions: move const version, better names in mod types

survived reviewed (approved):
Looks nice to me! But I don't have fresh eyes, I feel like I could miss something, I'd wait for @d86leader to have another pass on the PR

jkatzDfns commented on key-export-client/src/lib.rs:
```diff
@@ -0,0 +1,242 @@
+            .context("length of key_holders overflows u16")?;
+        if !(2 <= min_signers && min_signers <= n) {
+            return Err(new_error(
+                "threshold doesn't meet requirements: 2 <= min_signers <= n",
```
Requiring min_signers >= 2 seems a little ad hoc. Why not require min_signers >= 3? Or maybe in the future some customer who doesn't care about security (but does care about availability) does a 1-out-of-5 scheme?

OrestisAlpos replied:
I saw we had the same requirement in key-import, but probably this doesn't apply here.
Should I only keep "min_signers <= n" then @jkatzDfns ?

survived replied:
I think WASM module should accept any valid threshold, so it's $2 \le t \le n$. Any limitations that take into account the current deployment should be enforced by Coordinator and/or SDK

survived replied:
I assume that $t=1$ is not valid value for a threshold, because it defeats purpose of TSS by creating a SPOF at each signer. Also the math probably won't work well, for that reason $t=1$ is prohibited in cggmp21 and in the signers code.

jkatzDfns replied:
I did not realize that t=1 is prohibited in signers

jkatzDfns replied:
I guess it's fine to leave it, then. (FWIW, I still could imagine someday someone wants t=1 because they want to use our system but don't care about threshold.)

survived replied:
It means that signers share a polynomial of degree 0 (i.e. $F(x) = \text{sk}$). In the signing, threshold amount of signers (i.e. one) needs to be online. Protocol is not designed to support this case nor does it make any sense. Keygen - maybe, but signing should be different: each signer should do a regular signing when signature is requested (which is fast) instead of running CGGMP21 and trying to communicate to itself

survived replied:
I'd say we should push back that kind of requirement, because reducing SPOF is our security measure, and it's not compliant with decentralization vision of the future product (where no provider holds more than $t$ shares)

OrestisAlpos replied:
But it's only in the export functionality, I did not change the import.
Probably there will never exist a key with threshold 1 to export, for the reasons you mentioned, but is it a problem if the export code exports such keys as well?

survived replied:
I'm only saying that if we ever have $t=1$ keys, each "key share" would be just a key. It doesn't make sense to store "a key share" that contains a full key with a fields like other parties commitments (they are all the same) and shares indexes (they don't matter). Instead of key share, each signer would only store the full secret key. Export is trivial and does not require this WASM module, except maybe the encryption part.

jkatzDfns replied:
I don't feel strongly about it, and we could always change it later if we ever have a customer who wants threshold 1. I just don't see why we are enforcing anything. The coordinator tells us what t and n are, so what case would there be where the coordinator tells us some values of t and n and we say "no, those values are not allowed"?

jkatzDfns replied:
Not to mention the fact that currently we always have min_signers >= 3 ... so if we want error handling, then we should raise an error if that condition is ever violated

survived replied:
We should reject $t=1$ just because key export for those keys should be different

survived replied:
I agree that generally WASM should trust to coordinator on t,n values, but if WASM is called with $t=1$ then something wrong is going on

jkatzDfns replied:
ok

survived replied:
@OrestisAlpos can you change back to discarding $t=1$?

jkatzDfns commented on key-export-client/src/lib.rs:
```diff
@@ -0,0 +1,242 @@
+                    .context("interpolation failed")?
+            }
+            (_, _) => {
+                return Err(new_error("the combination of protocol and curve for this key is not supported for key export"));
```
The error should be more informative, i.e., saying that response.protocol using response.curve is not supported

jkatzDfns commented on key-export-client/src/lib.rs:
```diff
@@ -0,0 +1,242 @@
+        let secret_key = match (response.protocol, response.curve) {
+            (KeyProtocol::Cggmp21, KeyCurve::Secp256k1) => {
+                let key_shares = parse_key_shares::<Secp256k1>(&decrypted_key_shares)
+                    .context("cannot parse a decrypted secret-key share")?;
```
This error could also be more informative, i.e., indicating which signer's key share failed.

More generally, shouldn't it be possible to recover as long as min_signers of the shares can be parsed properly? Right now we fail even if only one key share is bad.

OrestisAlpos replied:
I had the impression that implicitly we assume all signers will behave correctly. But I will do what you suggested.

survived replied:
I'm not sure it's practical to make the code more complex to handle the cases when one of the signers send invalid key shares

survived replied:
@OrestisAlpos if this change noticeably complicates the code, let's discuss with @jkatzDfns whether we need to do that

OrestisAlpos replied:
No it doesn't complicate it that much. The new code will require some extra testing though.

jkatzDfns replied:
It's ok to fail if a share is incorrect, but the code should at least identify which signer's share caused the problem

jkatzDfns commented on key-export-client/src/lib.rs:
```diff
@@ -0,0 +1,242 @@
+) -> Result<SecretScalar<E>, InterpolateKeyError> {
+    // Validate input
+    let n = key_shares.len();
+    if n <= 1 {
```
If we disallow n=1 then this should be caught earlier (e.g., at line 111). But as per my point there, I am not sure we want to categorically exclude n=1

survived replied:
An extra validation of inputs never hurts...

jkatzDfns commented on key-export-client/src/lib.rs:
```diff
@@ -0,0 +1,242 @@
+                let public_key = Point::<Secp256k1>::from_bytes(&response.public_key)
+                    .context("cannot parse the public key")?;
+                interpolate_secret_key::<Secp256k1>(&key_shares, &public_key)
+                    .context("interpolation failed")?
```
Should probably give more information here about why interpolation failed, e.g., did it fail completely or did it succeed but the private key did not match the public key?

OrestisAlpos replied:
It actually does return this information. `context()` will display the given message and the one that comes from the lower-level error (in this case `interpolate_secret_key()` returns an enum `InterpolateKeyError`)

survived replied:
Yes, the error will be `interpolation failed: [stringifyied error returned by interpolate_secret_key]`. Possible errors returned by that function (and their stringified representation) can be found [here](https://github.com/dfns-labs/trusted-dealer/blob/823d894873de07feb4c0b72b54e06f6ff073db54/key-export-client/src/lib.rs#L218-L229)

jkatzDfns replied:
ok

jkatzDfns commented on key-export-client/src/lib.rs:
```diff
@@ -0,0 +1,242 @@
+                return Err(new_error("the combination of protocol and curve for this key is not supported for key export"));
+            }
+        };
+        Ok(SecretKey(secret_key))
```
Assuming everything works, do we return the private key? What format do we return it in?

OrestisAlpos replied:
We created a new type `SecretKey`, exposed to the JS user code, which has a function `to_bytes_be()` to get the content as a byte array. We decided not to return the bytes directly so that we can support other formats in the future (by implementing a new method for `SecretKey`)

jkatzDfns commented on key-export-client/tests/interpolate_secret_key.rs:
```diff
@@ -0,0 +1,95 @@
+    assert!(res.is_err());
+    assert!(matches!(
+        res.expect_err(""),
+        InterpolateKeyError::NotEnoughShares
```
Should this result in the same error as on line 62?

OrestisAlpos replied:
Yes, you are right. I will handle this together with your comment on the "n <= 1" condition.

survived replied:
@OrestisAlpos can you replace all the `.expect_err("")` with `.unwrap_err()`?

Commit:
more detailed error message; add test for recover_secret_key()

Commit:
change decryption and parsing of shares so as to tolerate errors, as long as a threshold of valid shares exists

Commit:
Previous commit changed interpolate_secret_key() to work with n > 0. Now added some tests for that case.

Commit:
change expect_err() to unwrap_err; :better comments

OrestisAlpos commented:
Ok, I have implemented some changes. 

The code now decrypts and parses the shares in a way that tolerates failures, as long as there at at least `min_signers` "valid". If this is not the case, it returns an error with the identities of the signers that sent an "invalid" (not decrypt-able or on parse-able) share.

The `recover_secret_key()` function can accept any threshold "1 <= min_signers && min_signers <= n". Technically there is indeed no way to require "2 <= min_signers" so I changed it, and we can let the coordinator/SDK enforce restrictions.

The `interpolate_secret_key()` function (called inside `recover_secret_key()` after decrypting and deserializing shares) can work with any positive number of shares --- if they are less than the threshold it will fail when comparing against the public key.

I also changed the error message for not supported protocol/curves.

survived commented on key-export-client/src/lib.rs:
```diff
@@ -108,53 +113,109 @@ impl KeyExportContext {
+        let error_msg = append_signer_ids_to_error_msg("not enough shares: the signers with the following idenities returned shares that cannot be decrypted: ".to_string(), &invalid_ids);
+        Err(new_error(&error_msg))
+    }
+}
```
> The code now decrypts and parses the shares in a way that tolerates failures, as long as there at at least min_signers "valid"

@OrestisAlpos @jkatzDfns I'm still not sure implementing the check was a good idea. What we're protecting from? Malicious signer? Malicious signer now can send a validly encrypted key share but with altered values, it can change secret share and/or commitments to make export fail. Unless we're doing key reconstruction with all possible combinations of `min_signers` key shares, we can't protect from this attack.

So why we're complicating the code?

jkatzDfns replied:
Isn't this redundancy one of the advantages of threshold? And it doesn't seem to complicate code too much. 

I agree it does not protect against a malicious signer, but it can correct errors that might occur at any point along the way.

OrestisAlpos replied:
Yes, the decrypted and parsed shares can still be invalid, and since by design we are only exporting the secret-key value (and not the commitments/validation data), we cannot check individual shares for validity.

The argument for keeping them would be that we protect from signers that follow the protocol (i.e., send their correct share) but are bootstrapped with invalid keys, or encode it wrong for some reason. Not sure how realistic this is to happen though. 
Plus, if we ever want, we can add the commitments and validate each share.

survived replied:
Commitments can't protect from an attack as commitments can be altered

survived replied:
> And it doesn't seem to complicate code too much.

Only this function was 8 lines and became 26 lines. It was easier and quicker to read. Now it requires effort to understand what is the logic behind the code.

survived replied:
It just doesn't seem to bring any practical value. I'd suggest to assume that ciphertexts don't accidentally get invalid on the way from the signer to the WASM module. And if they do, we should return error so client could go to us and we could investigate that.

jkatzDfns replied:
I have a hard time seeing the arguments against this. Sure, it's slightly more complex, but not much. And error in the ciphertexts seems more likely than getting t=1 from the coordinator.

survived replied:
I mean, yes, it may happen that ciphertexts get messed up. But if it happens, is it better to try to silently recover from this error, or return error and let it (eventually) trigger investigation?

survived replied:
If the ciphertexts are messed up, it (likely) can be caught at e2e tests before the release, as long as our module returns error instead of trying to recover

jkatzDfns replied:
ok, fair enough. But let's at least make sure we identify which share was corrupted to aid investigation

survived replied:
Yes that's definitely needed. @OrestisAlpos can you change decrypting and parsing back to what it was, but keep a nice errors generation?

OrestisAlpos replied:
So what do we do here?

survived replied:
We return error when one of the ciphertext cannot be decrypted or plaintext cannot be parsed, but make sure to include signer identity who sent this ciphertext

maurges commented on key-export-client/src/lib.rs:
```diff
@@ -0,0 +1,303 @@
+            ));
+        };
+
+        // Decrypt key shares
```
Nitpick: kind of unnecessary comment

maurges commented on key-export-client/src/lib.rs:
```diff
@@ -0,0 +1,303 @@
+/// Utility function to encode signer ids as Base64 and append them to a string
+fn append_signer_ids_to_error_msg(error_msg: String, invalid_ids: &[&Vec<u8>]) -> String {
+    let error_msg = invalid_ids.iter().fold(error_msg, |acc, id| {
+        acc + &general_purpose::STANDARD_NO_PAD.encode(id) + " "
```
Why not `acc + " " + encode(id)`? It seems at every usage place space is prepended

maurges commented on key-export-common/src/lib.rs:
```diff
@@ -0,0 +1,158 @@
+    ///Binance EDDSA
+    BinanceEcdsa,
+    ///CGGMP21
+    Cggmp21,
```
I'm burned by it in signers right now point: let's specify it instead as Cggmp21Secp256k1

survived replied:
This part matches [the spec](https://xkeyio.atlassian.net/wiki/spaces/EN/pages/209485842/Key+export)

survived replied:
You can communicate to @Thibault2ss if you want to change the spec, but I believe, it'd require to change their code as well at this point

jkatzDfns replied:
seems to make sense to have one enum for protocol and one enum for the curve, rather than a single enum covering both

maurges replied:
Ok, it seems that we do use them always together, so it's ok

Commit:
remove a comment; better error_msg format

Commit:
change field name to match specs

Commit:
example print request

Commit:
upd field name according to spec

Commit:
add js_name to wasm_bindgen

Commit:
camelCase in types consumed and returned by WASM code

Commit:
change KeyProtocol and KeyCurve serialization format to match the engineering definitions

Commit:
rename file as it now inlcuded more tests

Commit:
change decryption and parsing of key shares: return error when one of the ciphertexts cannot be decrypted or plaintext cannot be parsed, include signer identity who sent this ciphertext

Commit:
change back WASM recover_secret_key() function to disregard min_signers=1

OrestisAlpos commented:
I have implemented the latest changes

survived commented:
Should be good to merge at this point!
